providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -362,12 +395,13 @@ static int ml_kem_import(void *vkey, int selection, const OSSL_PARAM params[])
 
     include_private = selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY ? 1 : 0;
     res = ml_kem_key_fromdata(key, params, include_private);
-#ifdef FIPS_MODULE
     if (res > 0 && include_private && !ml_kem_pairwise_test(key)) {
+#ifdef FIPS_MODULE
         ossl_set_error_state(OSSL_SELF_TEST_TYPE_PCT);
+#endif
+        ossl_ml_kem_key_reset(key);
         res = 0;
     }
```

Comment:
Please put the test call inside the FIPS conditional:

```suggestion
#ifdef FIPS_MODULE
    if (res > 0 && include_private && !ml_kem_pairwise_test(key)) {
        ossl_set_error_state(OSSL_SELF_TEST_TYPE_PCT);
        ossl_ml_kem_key_reset(key);
        res = 0;
    }
#endif
```

The PCT is mandated by the FIPS 140-3 IGs but it's not providing anything useful and shouldn't be imposed on non-FIPS users of the algorithm.  Quite a few FIPS mandates are like this.

To be able to do a PCT, we need to have the private key.  What more does an attacker want when they've got that?  What benefit can be obtained by corrupting the key pair?  Corrupted key pairs are trivial to create, why go to all this effort?

I do agree that the verify call should do the PCT.  That's reasonable because our user is asking for it.  Doing it unconditionally on import isn't reasonable IMO.

Reply:
@openssl/committers thoughts?

Reply:
PCT is useful to detect corrupt/doctored keys in ML-KEM. The corrupt (bit flip, ...) case is more compelling of course.
In ML-KEM public keys cannot be fully derived from the (non-seed) private key, there's a random component that's not recoverable. This verifies that the private key and (embedded inside it) public actually match. It is cheap enough to do on private key import, which is not performance critical, and gives users a more meaningful early indication there's a problem with the key.

Reply:
To do a PCT you need both public and private keys.  They must also be loaded simultaneously.  This means that they are both coming from one source.  Manipulating bits in the public key isn't gaining anything because you've got the private key already.

All this will catch is accidental bit flips which will be caught later regardless.  I don't think imposing a performance penalty (and it is a penalty) on everyone just for the sake of something that will occur with minuscule probability is worthwhile.  I recollect that @t-j-h once said that he'd seen 1 or 2 random bit flips in all of his crypto experience.

Reply:
https://mailarchive.ietf.org/arch/browse/spasm/?gbt=1&index=5INXyopPkm_MAAiEvW_MufqGUKE

Reply:
I've only superficially looked at the paper.  I'll go through it in more detail next week.

The question that seems to need an answer: _do we want to try to detect a random bit flip in part of a key?_

This attack looks fairly theoretical.  It requires a random bit flip in a specific part of the key and then for an attacker to be able to detect this and make queries against it.  Bit flips are very rare, bit flips in the right place much rarer.  Being in a position where this can be exploited at the time it happens would also be extremely unlikely.  I fully expect that similar attacks could be made against any of the different types of asymmetric keys.

Specter (and similar) attacks are far more deterministic and much easier to exploit and we don't mitigate them.  I'm finding it hard to see this attack as plausible in practice.  There will be many random bit flip vulnerabilities in the code base (every conditional branch would qualify for a start).

This does highlight the fragility of the algorithm which is a concern.

Reply:
ML-KEM PCT is quite cheap, and note I am not adding it to keygen, where it remains conditional on FIPS, this is import only, so very much **not** performance impacting.
The objection appears to be perf-impact, which just isn't there.
Letting the user know there's a problem early before something goes south unexpectedly later is rarely objectionable.
So what is the actual problem here?
We've never done it before is not a strong argument.

Reply:
I must be missing something..
We have made the ML keys immutable so you either have a public key or a key pair.
The case you are suggesting would require you loading the public key into the key already containing a private key.

Reply:
No. The public key in question is the one embedded in the private key $`(sk = s|ek|H(ek)|z)`$.
The consistency check is on the parsed private key, which automatically carries a public component.

Reply:
Also, (something IIRC Shane was looking to see done) now check that explicit public keys match explicit private keys on import.
And cross-check the implicit rejection $`z`$ when both the seed and key are provided in a P8. Should probably do the same on import when both are given as parameters.

Reply:
If we used the seed in the import then doing a PCT seems really pointless. If we parsed the expanded key then I think it does make sense to do the PCT. This way if people use seed then they don't pay the performance penalty (regardless of the question of how big or small that penalty is).